core/src/main/java/ai/timefold/solver/core/impl/domain/entity/descriptor/EntityDescriptor.java

```diff
@@ -140,6 +140,10 @@ public class EntityDescriptor<Solution_> {
                 var moveDirector = (MoveDirector<Solution_, ?>) solutionView;
                 return !moveDirector.isPinned(this, entity);
             };
+    private final Map<EnumSet<ForEachFilteringCriteria>, Predicate<Object>> forEachIncludesToPredicateBasicVarMap =
+            new LinkedHashMap<>();
```

Comment:
I don't like seeing these complex types, especially when they're being used more than once. I recommend creating a new class this complexity; that class can then also be tested separately and isolated.